crates/sparrow-types/src/instantiate/type_constraints.rs

```diff
@@ -0,0 +1,169 @@
+///
+/// # Fails
+/// If the types are empty. If no parameters use the type class (and there are
+/// no associated parameters) we shouldn't try to determine the corresponding
+/// type.
+fn solve_type_class(parameter: &TypeParameter, types: &[FenlType]) -> Result<DataType, Error> {
+    debug_assert!(!types.is_empty());
+
+    debug_assert!(
+        parameter.type_classes.len() == 1,
+        "only one type class currently supported"
+    );
+    let type_class = parameter.type_classes[0];
+
+    // Find the minimum type compatible with all the arguments associated with the
+    // type class.
+    let mut result = types[0].clone();
+    for (index, arg_type) in types.iter().enumerate().skip(1) {
+        result = result.least_upper_bound(arg_type).ok_or_else(|| {
+            Error::IncompatibleTypesForConstraint {
+                type_parameter: parameter.name.clone(),
+                types: types[..=index].iter().cloned().collect(),
+            }
+        })?;
+    }
+
+    let result = match result.type_constructor {
+        TypeConstructor::Concrete(concrete) => concrete,
+        _ => panic!("Least upper bound should be concrete, but was {result:?}"),
+    };
+
+    let result = promote_concrete(result, type_class).ok_or_else(|| {
+        Error::IncompatibleTypesForConstraint {
+            type_parameter: parameter.name.clone(),
+            types: types.iter().cloned().collect(),
+        }
+    })?;
+
+    Ok(result)
+}
+
+/// Returns whether the `from` datatype can be implicitly cast to `to`.
+///
+/// For example, while it is possible to cast `i64` to `i32`, we don't
+/// want to do that implicitly since it "narrows" the type and discards
+/// values.
+fn can_implicitly_cast(from: &DataType, to: &DataType) -> bool {
+    use DataType::*;
+    match (from, to) {
+        (_, _) if from == to => true,
+        // Null can be cast to a null version of any other type.
+        (Null, _) => true,
+        // Integers can be widened or promoted to Float64.
+        (Int8, Int8 | Int16 | Int32 | Int64 | Float64) => true,
+        (Int16, Int16 | Int32 | Int64 | Float64) => true,
+        (Int32, Int32 | Int64 | Float64) => true,
+        (Int64, Int64 | Float64) => true,
+        // Unsigned integers can be widened or promoted to larger Int or Float64
+        (UInt8, UInt8 | UInt16 | UInt32 | UInt64 | Int16 | Int32 | Int64 | Float64) => true,
+        (UInt16, UInt16 | UInt32 | UInt64 | Int32 | Int64 | Float64) => true,
+        (UInt32, UInt32 | UInt64 | Int64 | Float64) => true,
+        (UInt64, UInt64 | Float64) => true,
+        // Floats can be widened.
+        (Float16, Float16 | Float32 | Float64) => true,
+        (Float32, Float32 | Float64) => true,
+        (Float64, Float64) => true,
+        // Other promotions that we allow implicitly.
+        (Utf8, Timestamp(TimeUnit::Nanosecond, None)) => true,
+        // Other promotions must be explicitly requested.
+        (_, _) => false,
+    }
```

Comment:
This is missing a few casts that are supported from the sparrow-compiler's version. On purpose?

Reply:
Unintentional -- probably reflects when I grabbed that code.

Reply:
Pulled in the latest.